Feat/pixiv deeplink

ilonachan opened the pull request:
Closes #9452.

The propsed implementation adds three new dummy search handlers:

- `aid:12345` for direct deeplinks to a specific illustration
- `sid:12345` for direct deeplinks to a specific series (both of these will be treated as "manga", and an illustration within the series points to the same manga as the series itself)
- `user:12345` to filter for only the illustrations by a specific user. This differs from the corresponding filter in that
  1. it is accessible to a deeplink
  2. it uses the user ID instead of the name
  3. it therefore works with users whose names contain spaces, unlike the filter variant

I chose these patterns so they hopefully won't be accidentally encountered by a user. I decided to make the last one more speaking, because it is potentially useful for regular searching (though its functionality could be expanded by allowing it to appear beside regular search terms). All of these decisions can obviously be debated, and would ideally be approved by someone who is more primarily responsible for this plugin.

It seems that the entire search mechanism of this plugin could benefit from a general overhaul (for example, to prevent the constant `IllegalArgumentException`s when "violating the [nonexistent] contract" by retrying a search). But without changing more than necessary, this is the best I can do so far.


Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [x] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

ilonachan committed:
Pixiv: added deeplink and ID search (#9452)

Direct ID search is triggered by prefixing the item's ID with with `aid:` for artworks/illustrations, `sid:` for series, and `user:` for users. The former two are meant only for use in deeplinks, while the latter may also be useful for actual users and therefore has a more exposable name. (All of these prefixes are subject to change)

ilonachan committed:
Pixiv: bandaid fix for API not returning needed values (#9452)

Apparently depending on the circumstances, the API doesn't return the user ID to which a series belongs. This user ID is instead placed in the outer Illustration object. This very basic (and subject to a larger refactoring) fix ensures that the user ID is always present when needed to construct the link (it isn't required for anything else in the API)

ilonachan committed:
Pixiv: ensured that only exact matches to the deeplink patterns are handled specially (#9452)

The exact pattern is: `<type>:<ID consisting of digits>`. By ensuring that only digits and nothing else afterwards are allowed by the pattern matching (otherwise falling back to regular search), we further decrease the likelihood of users accidentally triggering this functionality (it sadly can't be entirely avoided, since deeplinks need to share an interface with the regular search queries)

AwkwardPeak7 commented on src/all/pixiv/src/eu/kanade/tachiyomi/extension/all/pixiv/Pixiv.kt:
```diff
@@ -85,6 +88,25 @@ class Pixiv(override val lang: String) : HttpSource() {
+        // Deeplink selection of specific IDs: simply fetch the single object and return
+        if (query.startsWith(ID_ILLUST_PREFIX) &&
+            query.removePrefix(ID_ILLUST_PREFIX).matches("\\d+".toRegex())
+        ) {
```
```suggestion
        if (query.startsWith(ID_ILLUST_PREFIX)) {
```
no need to check, since anything else will result in 404 or similar from pixiv

ilonachan replied:
The reason I added that check is so regular queries that just happen to look like `sid:dharta or sth` aren't captured by this mechanism. Idea being that the user either knows about this special handling, or must be *extremely* unlucky to trigger it. But if you don't think it's important to go that far, yeah we can skip the check.

ilonachan replied:
I've decided (for now) to move completely to the URL-based system, meaning this isn't really relevant anymore. I've left the code for parsing and handling the special search queries in, including this check, because especially the way they're now conceptually set up there needs to be *some* validation. But it's currently not doing anything anyway, and the exact way it's gonna be used/useful/changed/deleted in the future is completely up in the air for now.

AwkwardPeak7 commented on src/all/pixiv/src/eu/kanade/tachiyomi/extension/all/pixiv/Pixiv.kt:
```diff
@@ -85,6 +88,25 @@ class Pixiv(override val lang: String) : HttpSource() {
+        }
+        if (query.startsWith(ID_SERIES_PREFIX) &&
+            query.removePrefix(ID_SERIES_PREFIX).matches("\\d+".toRegex())
+        ) {
```
same here

ilonachan replied:
(see [above](https://github.com/keiyoushi/extensions-source/pull/9457#discussion_r2173485971))

AwkwardPeak7 commented on src/all/pixiv/src/eu/kanade/tachiyomi/extension/all/pixiv/Pixiv.kt:
```diff
@@ -95,19 +117,35 @@ class Pixiv(override val lang: String) : HttpSource() {
-                    filters.makeUsersPredicate()?.let(::add)
+                // TODO: it would be useful to allow multiple user: tags in the query
+                if (query.startsWith(ID_USER_PREFIX) &&
+                    query.removePrefix(ID_USER_PREFIX).matches("\\d+".toRegex())
```
same here

ilonachan replied:
(see [above](https://github.com/keiyoushi/extensions-source/pull/9457#discussion_r2173485971))

AwkwardPeak7 commented on src/all/pixiv/src/eu/kanade/tachiyomi/extension/all/pixiv/PixivConstants.kt:
```diff
@@ -0,0 +1,7 @@
+package eu.kanade.tachiyomi.extension.all.pixiv
+
+object PixivConstants {
```
no need for separate object, global constants should be fine

ilonachan replied:
Oh yeah right this is Kotlin :rofl: My own first instinct is that having constants just... be there, floating around in the package, feels kinda weird. For super generally useful ones maybe, but the instinct to put a static holder class around them is strong. But I imagine that's not the kotlin-esque way to do it, so sure I can get rid of the wrapper.

AwkwardPeak7 commented on src/all/pixiv/src/eu/kanade/tachiyomi/extension/all/pixiv/PixivUrlActivity.kt:
```diff
@@ -0,0 +1,70 @@
+        do {
+            var pathSegments = intent?.data?.pathSegments ?: break
+
+            if ("en".equals(pathSegments[0])) {
```
en should not be hardcoded? since pixiv supports other languages

ilonachan replied:
I would've thought so too, but every language code I've tried just gives a 404 (even "jp"). So for now I assumed that "en" is special for some weird reason.

But on the off chance that there are other codes that would work (or you or someone else knows them already), what should I do instead? Make a whitelist of those codes, or just allow trying any known language code? (this might balloon the `intent-filter` list tho.) Or another wildcard? (that might complicate matching logic a bit.)

ilonachan replied:
For now I've gone with a `KNOWN_LOCALES` constant, a list with (currently) only one entry. I don't understand the entirety of Pixiv's URL system, so until someone with more concrete knowledge can clarify things this is probably the best solution for now.

AwkwardPeak7 commented on src/all/pixiv/AndroidManifest.xml:
```diff
@@ -0,0 +1,29 @@
+                <data android:scheme="http" />
+                <data android:scheme="https" />
+
+                <data android:pathPattern="/en/artworks/..*" />
```
en shouldn't be strict requirement? since pixiv supports other languages

ilonachan committed:
Pixiv: changed Deeplink system to use URL

Instead of complex parsing logic in the (deliberately lightweight and kotlin-wise handicapped) Deeplink Activity, the captured URL can just be passed to the search directly and handled there. The ability for the search to understand full Pixiv URLs is useful (and half-expected) either way, and it will not interfere with regular search function.

ilonachan committed:
Pixiv: fixed IndexOOB when query is empty/not a valid URI

AwkwardPeak7 commented on src/all/pixiv/src/eu/kanade/tachiyomi/extension/all/pixiv/PixivLinkUtil.kt:
```diff
@@ -0,0 +1,101 @@
+
+sealed class PixivTarget {
+    companion object {
+        val BASE_URI = Uri.parse("https://www.pixiv.net")
```
consider using okhttp's `HttpUrl` instead of `Uri`

ilonachan replied:
I mostly used `Uri` because of Android compatibility, the intent gives us a `Uri` after all. But maybe `HttpUrl` should be the default, and `from(uri: Uri)` could just be an overload. I assume the only difference is that okhttp actually does validation?

AwkwardPeak7 commented on src/all/pixiv/src/eu/kanade/tachiyomi/extension/all/pixiv/PixivLinkUtil.kt:
```diff
@@ -0,0 +1,101 @@
+
+        override fun toUri(): Uri {
+            if (authorUserId == null) {
+                TODO("Not yet implemented")
```
throw regular exception, cuz iirc this can crash the app

ilonachan committed:
Pixiv: Applied suggestion to use OkHttp Urls
